### PR TITLE
Introducing default mime type for metadata.

### PR DESCRIPTION
**Pull request by NiteshKant**
As per discussion in issue #89, this change proposes a default mime-type for metadata payloads.

Existing `Schema.md` is replaced by this new `MimeTypes.md` document which elaborates this new mime type.

This proposal is to specify a default mime type for metadata payload. In absence of a user specifying a mime type (while creating a reactive socket), this default will be used. This does _not_ make the metadata mime type field in the setup frame as optional.

### CODE REVIEW

File: MimeTypes.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+## Mime types
+
+A [Setup Frame](https://github.com/ReactiveSocket/reactivesocket/blob/master/Protocol.md#setup-frame) in the Reactive
+Socket protocol specifies mime types for both [data and metadata](https://github.com/ReactiveSocket/reactivesocket/blob/master/Protocol.md#data-and-metadata) payloads of a frame.
```
Comment (stevegury) [inline MimeTypes.md]: I think it's better to avoid breaking up the lines explicitly.

File: MimeTypes.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+This document provides a guidance for the mime types of these payloads.
+
+#### Metadata
+
+Reactive Socket provides a default mime type for metadata payloads and all implementations of the protocol MUST use
```
Comment (stevegury) [inline MimeTypes.md]: ReactiveSocket in one word.

File: MimeTypes.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+##### Overview
+
+The default mime type assumes key-value metadata fields with the following encodings:
+
+* __Key__   : The value is always a UTF-8 string.
+* __Value__ : Value is an opaque binary value and the encoding can vary from one field to another.
+
+##### Name
+
+The name of the default mime type for metadata payloads is __application/x.reactivesocket+kv__. This name MUST be
```
Comment (stevegury) [inline MimeTypes.md]: Why the "__" before/after?

**Comment by stevegury** [comment]
Other than those 2 comments, LGTM 👍

**Comment by benjchristensen** [comment]
Curious, why define our own encoding format as opposed to adopting an industry standard?

**Comment by benjchristensen** [comment]
> In absence of a user specifying a mime type (while creating a reactive socket), this default will be used. This does not make the metadata mime type field in the setup frame as optional.

Those sentences seem contradictory. How would it not be optional if "in absence of a user specifying" is an option?

**Comment by NiteshKant** [comment]
(Combining responses into a single comment)

@stevegury I will remove the line breaks and use `ReactiveSocket` as a single word. `__` is for making the word bold.

@benjchristensen 
>> Curious, why define our own encoding format as opposed to adopting an industry standard?

The format here is pretty simple with keys as only strings and values as opaque bytes payload. It assumes that if the value is a complex type, it will use some standard format. Most of the standards are pretty heavyweight for this simple requirement. Having said that, if you have something in mind, let me know and I will look at it.

>> Those sentences seem contradictory. How would it not be optional if "in absence of a user specifying" is an option?

I think the language can be improved. Here is what it means:

- The SETUP frame __requires__ a mime type for metadata. This proposal does not relax that requirement i.e. you can not send a SETUP frame without a metadata mime type.
- Any implementation (like the java impl here) MAY provide a way to create a reactive socket without providing a mime type. In such a case, this default mime type should be used.

**Comment by benjchristensen** [comment]
> The format here is pretty simple with keys as only strings and values as opaque bytes payload. It assumes that if the value is a complex type, it will use some standard format. Most of the standards are pretty heavyweight for this simple requirement. Having said that, if you have something in mind, let me know and I will look at it.

My concern is that by reinventing the "simple" approach, are we putting ourselves at risk of repeating what HTTP did with headers in v1 (https://http2.github.io/http2-spec/compression.html#rfc.section.1)?

**Comment by stevegury** [comment]
The difference with HTTP/1 is that with HTTP/1 you __CAN'T__ change the metadata format. Here you still can change the format, we just provide one for free.

**Comment by NiteshKant** [comment]
@benjchristensen I agree it is reinventing a concept that is common for most protocols and HTTP/2 has done a lot of work to "undo" what HTTP/1 did. Lot of HTTP/2's complexity in header compression comes from the fact that it is backward compatible with HTTP/1 and the fact that user agents can keep pretending that they are doing HTTP/1. 
Having said the above, one of the points in the [original PR](https://github.com/ReactiveSocket/reactivesocket/issues/89) was to see whether it is worth plain forklifting HTTP/2 hpack since it is a standard. We can discuss that now if required.

Although, this mime-type is optional, I would like to do the due diligence that it at least works for most cases. If not, there is less value in providing a default that many won't use.

**Comment by tmontgomery** [comment]
Providing a default encoding is probably useful. But I agree with @benjchristensen that it might be best to base it on a standard (or as close as possible). Would a streamlined CBOR or CBOR-compatible encoding work here?

**Comment by NiteshKant** [comment]
@tmontgomery there are only the following data types to be encoded here:

- Map (Implicit as it is a list of K-V pairs)
- UTF-8 string (key)
- Opaque binary payload (value)

I believe the CBOR representation will be pretty close to what is present here. This representation is a bit more compact as we know the types before hand and we just need the lengths. IMHO, picking up CBOR does not provide much value in this case but I am open to it if we have a consensus here that it is better to choose that.

Personally, I believe there is more value in picking up a standard which solves similar problem i.e. representation of headers/metadata.

**Comment by tmontgomery** [comment]
@NiteshKant using the CBOR encoding for those 3 pieces is what I would recommend. It's a standardized (as much as anything) encoding. Can be used with multiple implementations. And has multiple language support already.

Full on CBOR encoding isn't necessary as you point out. However, using a standard encoding that is pretty lightweight, can be referred to, and has multiple implementation support is useful, I think.

By contrast, a specific, new encoding for this makes it a one off that implementations would have to support. Being simple is good, though. But a specific subset of CBOR is simple and a standard. ;)

I would suggest HPACK, but I don't know of any easy way to use a subset of HPACK that is _relatively_ simple.

**Comment by NiteshKant** [comment]
Ok great I am good with CBOR then. 

So, I guess we would have to specify in the spec as to what will be the structure of the metadata payload (map of UTF-8 strings to binary values)? Or we just leave it open and say we use CBOR?

**Comment by stevegury** [comment]
I think using CBOR is fine but I don't like forcing the user to have a dependency to a CBOR library (this is why I proposed a custom implementation).
Maybe this default implementation could be moved to another jar (something like reactivesocket-default-encoding)?

**Comment by NiteshKant** [comment]
@stevegury So, from the java implementation point of view, we have two option:

1. Make [reactivesocket-java](https://github.com/ReactiveSocket/reactivesocket-java) project multi-module and add `reactivesocket-default-encoding` module.
1. Add the `reactivesocket-default-encoding` module to [reactivesocket-java-impl](https://github.com/ReactiveSocket/reactivesocket-java-impl) project.

I would prefer 2 above. Thoughts?

**Comment by stevegury** [comment]
I prefer 2 too.

**Comment by benjchristensen** [comment]
> I don't like forcing the user to have a dependency

Agreed. 

> Add the reactivesocket-default-encoding module to reactivesocket-java-impl project

I prefer this as well if we are providing a default implementation of the default encoding.

File: Protocol.md [hunk 4]
```diff
@@ -571,7 +571,7 @@ The general format for an extension frame is given below.
      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
     |R|                    Frame Length (optional)                  |
-    +-------------------------------+-+-+---------------------------+
+    +-+------------------------------+-+-+--------------------------+
```
Comment (yschimke) [inline Protocol.md]: Are the 3 in the middle out of place now?

Comment (NiteshKant) [inline Protocol.md]: Yeah 😢  I will revert this change as the missing `+` is for all frames.

**Comment by NiteshKant** [comment]
So, this is ready for another review. 

The implementation is in this PR:
https://github.com/ReactiveSocket/reactivesocket-java-impl/pull/11

**Comment by NiteshKant** [comment]
If there aren't any objections, I would like to merge this. 
Implementation is merged in the java-impl project.

**Pull request merged**